### PR TITLE
Shallow clean-up of the solidity code

### PR DESCRIPTION
**Pull request by AntoineRondelet**
Related to: https://github.com/clearmatics/zeth/issues/321 and https://github.com/clearmatics/zeth/issues/192

This PR:
- Adjusts the Zeth solidity coding standards to rather align with the "language standards" proposed by the community and the solidity team. The driving factors for that are:
    - If there is a standard we should do our best to follow it (good practice) - it removes friction as to "what looks better" and just follows the norm.
    - Removes sources of inconsistencies between the code we write and the code we import (which catches the eye pretty badly when we call a function from an imported contract, like an ERC token for instance)
- Renames all solidity constants in `UPPER_SNAKE_CASE`
- Adds explicit scopes to all functions and storage variables (this should become systematic from now on as more verbosity with regard to the scopes can only be better)
- Switches the linter to `solhint` (`solium` then replaced by `ethlint` seems to be poorly maintained, last release was more than a year ago. `solhint` is actively maintained offers a good configuration set and is used by openZeppelin (https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/.solhint.json) and other serious projects). Several checks are voluntarily disabled for now because I didn't want to end up with a PR that shuffles the entire code base (renaming public functions, contract names etc => cascading these changes to the python code etc which will become a nightmare to review. Let's take it incrementally to avoid adding mistakes to the code base and breaking things)
- Uses `NatSpec` for documentation comments
- Orders functions (as per: https://docs.soliditylang.org/en/develop/style-guide.html#order-of-functions)
- Fixes the order of the modifiers

TODO:
- [x] Fix the `test_merkle_tree_contract.py` test (we have a `MerkleTreeMiMC7_test` test contract but its import is broken since `MerkleTreeMiMC7.sol` does not exist anymore)
- [x] Fix the `test_contract_base_mixer.py` test. It is now broken because the scope of some functions (e.g. `test_assemble_nullifiers`) has been changed in this PR, so these functions are not in the ABI anymore. Since we should not expose unnecessary function simply to be able to call them in the test scripts, let's add a `MixerBase_test.sol` "test proxy contract" to re-enable the tests despite the functions visibility change. - Fixed in #338

### CODE REVIEW

File: zeth_contracts/contracts/BLS12_377_test.sol [hunk 1]
```diff
@@ -9,39 +9,39 @@ contract BLS12_377_test
...
+    uint256 internal constant SCALAR_BYTES = SCALAR_WORDS * 32; // 32 (0x20)
 
-    uint256 constant g1CoordWords = 2;
-    uint256 constant g1CoordBytes = g1CoordWords * 32; // 64 (0x40)
-    uint256 constant g1PointWords = 2 * g1CoordWords;
-    uint256 constant g1PointBytes = g1PointWords * 32; // 128 (0x80)
+    uint256 internal constant G1_COORD_WORDS = 2;
+    uint256 internal constant G1_COORD_BYTES = G1_COORD_WORDS * 32; // 64 (0x40)
+    uint256 internal constant G1_POINT_WORDS = 2 * G1_COORD_WORDS;
+    uint256 internal constant G1_POINT_BYTES = G1_POINT_WORDS * 32; // 128 (0x80)
```
Comment (AntoineRondelet) [inline zeth_contracts/contracts/BLS12_377_test.sol:18]: Unused?

File: zeth_contracts/contracts/BLS12_377_test.sol [hunk 2]
```diff
@@ -9,39 +9,39 @@ contract BLS12_377_test
...
-    uint256 constant scalarBytes = scalarWords * 32; // 32 (0x20)
+    uint256 internal constant SCALAR_WORDS = 1;
+    uint256 internal constant SCALAR_BYTES = SCALAR_WORDS * 32; // 32 (0x20)
 
-    uint256 constant g1CoordWords = 2;
-    uint256 constant g1CoordBytes = g1CoordWords * 32; // 64 (0x40)
-    uint256 constant g1PointWords = 2 * g1CoordWords;
-    uint256 constant g1PointBytes = g1PointWords * 32; // 128 (0x80)
+    uint256 internal constant G1_COORD_WORDS = 2;
+    uint256 internal constant G1_COORD_BYTES = G1_COORD_WORDS * 32; // 64 (0x40)
```
Comment (AntoineRondelet) [inline zeth_contracts/contracts/BLS12_377_test.sol:16]: Unused?

File: zeth_contracts/contracts/BLS12_377_test.sol [hunk 3]
```diff
@@ -9,39 +9,39 @@ contract BLS12_377_test
     // In many cases, these numbers must be used as literals in the assembly
     // code.
 
-    uint256 constant scalarWords = 1;
-    uint256 constant scalarBytes = scalarWords * 32; // 32 (0x20)
+    uint256 internal constant SCALAR_WORDS = 1;
+    uint256 internal constant SCALAR_BYTES = SCALAR_WORDS * 32; // 32 (0x20)
```
Comment (AntoineRondelet) [inline zeth_contracts/contracts/BLS12_377_test.sol:13]: Unused?

File: zeth_contracts/contracts/BLS12_377_test.sol [hunk 4]
```diff
@@ -9,39 +9,39 @@ contract BLS12_377_test
...
+    uint256 internal constant G1_POINT_BYTES = G1_POINT_WORDS * 32; // 128 (0x80)
 
-    uint256 constant g2CoordWords = 4;
-    uint256 constant g2CoordBytes = g2CoordWords * 32; // 128 (0x80)
-    uint256 constant g2PointWords = 2 * g2CoordWords;
-    uint256 constant g2PointBytes = g2PointWords * 32; // 256 (0x100)
+    uint256 internal constant G2_COORD_WORDS = 4;
+    uint256 internal constant G2_COORD_BYTES = G2_COORD_WORDS * 32; // 128 (0x80)
+    uint256 internal constant G2_POINT_WORDS = 2 * G2_COORD_WORDS;
+    uint256 internal constant G2_POINT_BYTES = G2_POINT_WORDS * 32; // 256 (0x100)
```
Comment (AntoineRondelet) [inline zeth_contracts/contracts/BLS12_377_test.sol:23]: Unused?

File: zeth_contracts/contracts/BLS12_377_test.sol [hunk 5]
```diff
@@ -9,39 +9,39 @@ contract BLS12_377_test
...
+    uint256 internal constant G1_COORD_BYTES = G1_COORD_WORDS * 32; // 64 (0x40)
+    uint256 internal constant G1_POINT_WORDS = 2 * G1_COORD_WORDS;
+    uint256 internal constant G1_POINT_BYTES = G1_POINT_WORDS * 32; // 128 (0x80)
 
-    uint256 constant g2CoordWords = 4;
-    uint256 constant g2CoordBytes = g2CoordWords * 32; // 128 (0x80)
-    uint256 constant g2PointWords = 2 * g2CoordWords;
-    uint256 constant g2PointBytes = g2PointWords * 32; // 256 (0x100)
+    uint256 internal constant G2_COORD_WORDS = 4;
+    uint256 internal constant G2_COORD_BYTES = G2_COORD_WORDS * 32; // 128 (0x80)
```
Comment (AntoineRondelet) [inline zeth_contracts/contracts/BLS12_377_test.sol:21]: Unused?

File: zeth_contracts/contracts/BW6_761_test.sol [hunk 6]
```diff
@@ -9,18 +9,18 @@ contract BW6_761_test
...
     // code.
 
-    uint256 constant scalarWords = 2;
-    uint256 constant scalarBytes = scalarWords * 32; // 64 (0x40)
-    uint256 constant coordWords = 3;
-    uint256 constant coordBytes = 3 * 32; // 96 (0x60)
-    uint256 constant pointWords = 2 * coordWords;
-    uint256 constant pointBytes = pointWords * 32; // 192 (0xc0)
+    uint256 internal constant SCALAR_WORDS = 2;
+    uint256 internal constant SCALAR_BYTES = SCALAR_WORDS * 32; // 64 (0x40)
```
Comment (AntoineRondelet) [inline zeth_contracts/contracts/BW6_761_test.sol:13]: Unused?

File: zeth_contracts/contracts/BW6_761_test.sol [hunk 7]
```diff
@@ -9,18 +9,18 @@ contract BW6_761_test
...
+    uint256 internal constant COORD_WORDS = 3;
+    uint256 internal constant COORD_BYTES = 3 * 32; // 96 (0x60)
```
Comment (AntoineRondelet) [inline zeth_contracts/contracts/BW6_761_test.sol:15]: Unused?

File: zeth_contracts/contracts/BW6_761_test.sol [hunk 8]
```diff
@@ -9,18 +9,18 @@ contract BW6_761_test
...
+    uint256 internal constant POINT_WORDS = 2 * COORD_WORDS;
+    uint256 internal constant POINT_BYTES = POINT_WORDS * 32; // 192 (0xc0)
```
Comment (AntoineRondelet) [inline zeth_contracts/contracts/BW6_761_test.sol:17]: Unused?

**Comment by AntoineRondelet** [comment]
@dtebbs there are some solidity variables that are seemingly unused. Could you double check this before we go further and consider removing them? (Are you using them somewhere I missed?)

File: zeth_contracts/contracts/BW6_761_test.sol (hunk 8 above)
Comment (AntoineRondelet) [reply zeth_contracts/contracts/BW6_761_test.sol:17]: Btw all these can be made `private` I think. Would be good to tighten the scopes.

**Comment by dtebbs** [comment]
> @dtebbs there are some solidity variables that are seemingly unused. Could you double check this before we go further and consider removing them? (Are you using them somewhere I missed?)

@AntoineRondelet  Yes, they do seem to unused.  Good catch.

File: zeth_contracts/.solhint.json [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+        "func-name-mixedcase": "off",
+        "contract-name-camelcase": "off",
+        "no-empty-blocks": "off",
+        "compiler-version": "off",
+        "reason-string": "off",
+        "avoid-suicide": "error",
+        "no-inline-assembly": "off",
+        "avoid-sha3": "warn",
+        "func-visibility": "error"
+    }
```
Comment (dtebbs) [inline zeth_contracts/.solhint.json:15]: Just FYI, I tried adding `"max-line-length": ["error",80]`, but this catches comments (e.g. long URLs).  Would be nice to have that at some point, though, perhaps using `// solhint-diable-next-line ...` where long lines are unavoidable.

File: zeth_contracts/package.json [hunk 10]
```diff
@@ -10,16 +10,16 @@
     "deploy": "truffle deploy",
     "test": "truffle test",
     "debug": "truffle debug",
-    "check": "node_modules/solium/bin/solium.js --dir contracts"
+    "check": "node_modules/solhint/solhint.js 'contracts/**/*.sol'"
```
Comment (dtebbs) [inline zeth_contracts/package.json:13]: This command does not exit with error for warnings (I tried changing a constant to lower case).  This means that the CI will never detect these issues.

**Review by dtebbs — COMMENTED** [review]
Looks good.  Would be nice to have the CI catch violations.  (I'll look into that)

**Comment by dtebbs** [comment]
I've tried to address my comments in #334

File: zeth_contracts/.solhint.json (hunk 9 above)
Comment (AntoineRondelet) [reply zeth_contracts/.solhint.json:16]: Thanks #334 was merged in this branch.

**Comment by AntoineRondelet** [comment]
Note: the `test_contract_base_mixer.py` test is broken. To properly fix it we would need to use a "test proxy contract", but this is for later (see #335). For now, let's just roll back and set the scopes of the tested functions back to `public` and update the tests to make sure that the tested functions behave correctly.

**Comment by dtebbs** [comment]
(Worth rebasing this onto origin/dvelop before merging)

**Comment by dtebbs** [comment]
#338 fixes the tests and provides public wrappers around the internal functions (so no need to adjust this code)

**Pull request merged**